pkg/cdm/bootstrap.go

```diff
@@ -0,0 +1,353 @@
+				Gateway: config.ManagementGateway,
+			},
+		}
+	}
+
+	endpoint := "/cluster/me/bootstrap"
+	buf, code, err := c.post(ctx, nodeIP, endpoint, Internal, bootstrapConfig)
+	if err != nil {
+		return 0, fmt.Errorf("failed POST request %q: %s", endpoint, err)
+	}
+	if code != 202 {
+		return 0, fmt.Errorf("failed POST request %q: %s", endpoint, http.StatusText(code))
+	}
+
+	var bootstrap struct {
+		ID int `json:"id"`
+	}
+	if err := json.Unmarshal(buf, &bootstrap); err != nil {
+		return 0, fmt.Errorf("failed to unmarshal bootstrap status: %s", err)
+	}
+
+	return bootstrap.ID, nil
+}
+
+// IsBootstrapped returns true if the cluster has been bootstrapped, false
+// otherwise. The cluster can be rebooted at any time when this function runs,
+// the timeout parameter controls how long we wait for the cluster to become
+// responsive again.
+func (c *BootstrapClient) IsBootstrapped(ctx context.Context, nodeIP string, timeout time.Duration) (bool, error) {
+	c.Log.Print(log.Trace)
+
+	var failure time.Time
+	for {
+		buf, code, err := c.get(ctx, nodeIP, "/node_management/is_bootstrapped", Internal)
+		if err == nil && code == 200 {
+			var isBootstrapped struct {
+				Value bool `json:"value"`
+			}
+			if err := json.Unmarshal(buf, &isBootstrapped); err != nil {
+				return false, fmt.Errorf("failed to unmarshal bootstrap status: %s", err)
+			}
+
+			return isBootstrapped.Value, nil
+		}
+
+		if errors.Is(err, context.DeadlineExceeded) {
+			return false, err
+		}
+		if err == nil {
+			err = errors.New(http.StatusText(code))
+		}
+		if failure.IsZero() {
+			failure = time.Now()
+		}
+		if !failure.IsZero() && time.Since(failure) > timeout {
+			return false, err
+		}
+
+		c.Log.Printf(log.Debug, "Request returned: %s, retrying", err)
+		time.Sleep(defaultWait)
+	}
+}
+
+// WaitForBootstrap blocks until the bootstrapping of the cluster succeeds or
+// fails. The cluster can be rebooted at any time when this function runs, the
+// timeout parameter controls how long we wait for the cluster to become
+// responsive again.
+func (c *BootstrapClient) WaitForBootstrap(ctx context.Context, nodeIP string, requestID int, timeout time.Duration) error {
+	c.Log.Print(log.Trace)
+
+	var failure time.Time
+	for {
+		res, code, err := c.get(ctx, nodeIP, fmt.Sprintf("/cluster/me/bootstrap?request_id=%d", requestID), Internal)
+		if err == nil && code == 200 {
+			failure = time.Time{}
+
+			var bootstrap struct {
+				Status  string `json:"status"`
+				Message string `json:"message"`
+			}
+			if err := json.Unmarshal(res, &bootstrap); err != nil {
+				return fmt.Errorf("failed to unmarshal bootstrap status: %s", err)
+			}
+			switch bootstrap.Status {
+			case "IN_PROGRESS":
+				c.Log.Print(log.Debug, "Bootstrap in progress")
+				time.Sleep(defaultWait)
+				continue
+			case "FAILURE", "FAILED":
+				return fmt.Errorf("bootstrap failed: %s", bootstrap.Message)
+			default:
+				return nil
+			}
+		}
+
+		if errors.Is(err, context.DeadlineExceeded) {
```

Comment:
Wouldn't it be better to check `ctx.Err()` here, otherwise we won't be able to catch if the parent calls `cancel()`?

Reply:
👍